### PR TITLE
[SPARK-10387][ML] Add code gen for gbt

### PR DESCRIPTION
**Pull request by holdenk**
This PR adds code gen for GBT in ML. It does this using quasi quotes since they give nice compile time type messages. The API could also use some feedback for how we want to expose code gen - the current iteration gives the user a `toCodeGen` function they can call, but we could also use a config flag or something similar (open to suggestions). This is based on DB's work (as mentioned in the Jira)

### CODE REVIEW

**Comment by rxin** [comment]
Please don't use macros and quasiquotes. They are slow and not IDE friendly. We replaced them with Janino in SQL.

**Comment by holdenk** [comment]
@rxin Ok I'll update this to use Janino although I'll miss my compile time type checking I figured I'd probably have to do that. In the meantime does the API look like an OK place to start.

**Comment by rxin** [comment]
there isn't really compile time type checking -- it just has syntax checking to ensure it's a valid tree...

**Comment by holdenk** [comment]
ah ok - well less of a loss then. I'll switch it around of course.

**Comment by holdenk** [comment]
streaming test failure seems unrelated but won't retrigger since need to rewrite the codegen anyways.

**Comment by holdenk** [comment]
jenkins, retest this please

**Comment by holdenk** [comment]
I think the mesos test failure was unrelated, jenkins retest this please.

**Comment by holdenk** [comment]
So @rxin I reworked it to use java codegen - does it look ok?

**Comment by holdenk** [comment]
Also ping @dbtsai since I think we've branched 1.6 this should be in an ok state to review.

**Comment by holdenk** [comment]
I discovered a possible issue during some more thorough performance benchmarking - switching this to WIP while I investigate.

**Comment by rxin** [comment]
Mind closing and reopening when it is ready?

**Comment by holdenk** [comment]
Sure thing.

**Comment by dbtsai** [comment]
@holdenk Ping me when you address the issue and reopen it.

**Comment by holdenk** [comment]
@dbtsai : I've fixed the issue with generating large trees (depth > 10) that I found while doing the performance testing.

File: mllib/pom.xml [hunk 1]
```diff
@@ -125,6 +125,10 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-test-tags_${scala.binary.version}</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.codehaus.janino</groupId>
+      <artifactId>janino</artifactId>
+    </dependency>
```
Comment (dbtsai) [inline mllib/pom.xml]: I thought `sql` already has `org.codehaus` dependencies. Why do we need to add it here?

Comment (holdenk) [inline mllib/pom.xml]: I thought it would be better to explicitly depend on it now that we are using rather than depending on a transitive inclusion from another component.

File: mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+          case _ => {
+            val nodeSplit = node.split
+            val (leftSubCode, leftSubFunction) = nodeToTree(node.leftChild, depth + 1)
+            val (rightSubCode, rightSubFunction) = nodeToTree(node.rightChild, depth + 1)
+            val subCode = nodeSplit match {
+              case split: CategoricalSplit => {
+                val isLeft = split.isLeft
+                isLeft match {
+                  case true => s"""
+                              if (categories.contains(input.apply(${split.featureIndex}))) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala]: Why code-gen is in the way? Supposedly, after code-gen, all the condition value should be in the `if-else` statement. Here, `input.apply(${split.featureIndex}) == somevalue`

File: mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                } else {
+                 ${rightSubCode}
+                }"""
+              }
+            }
+            (subCode, leftSubFunction + rightSubFunction)
+          }
+        }
+      }
+      case node: LeafNode => (s"return ${node.prediction};", "")
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala]: Do you actually flat the tree scoring? Seems that between node to node, it's still recursion. We should have a code-gen for the entire tree.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala]: oh.. my bad. i got ur code.

File: mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala [hunk 4]
```diff
@@ -85,7 +85,7 @@ final class CategoricalSplit private[ml] (
   /**
    * If true, then "categories" is the set of categories for splitting to the left, and vice versa.
    */
-  private val isLeft: Boolean = _leftCategories.length <= numCategories / 2
+  private[ml] val isLeft: Boolean = _leftCategories.length <= numCategories / 2
 
   /** Set of categories determining the splitting rule, along with [[isLeft]]. */
   private val categories: Set[Double] = {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala]: For this, I will use BitSet with Int.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala]: Should we maybe do this in a second change? It seems like if there aren't that many categories for one direction then this would maybe not be a great change. What are your thoughts?

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala]: We use `BitSet` in our system. I know `BitSet` takes fast constant, while `HashSet` takes effectively longer constant time depending on the length of elements. Also, `HashSet` will box the element, and `BitSet` stores as primitive elements. But I don't benchmark myself. Agree, this can be a separate PR.

**Comment by holdenk** [comment]
So I changed this a bit @dbtsai (mostly just avoid wrapping the call in one place) - let me know what you think.

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 5]
```diff
@@ -189,6 +190,18 @@ final class GBTClassificationModel private[ml](
 
   override def treeWeights: Array[Double] = _treeWeights
 
+  // We use two vals with options rather than lazy vals since we want
+  // the codeGen to be eagerly evaluated.
+  val treePredictors = useCodeGen match {
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: Is there a reason to prefer match over if/else here? If/else seems more intuitive.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: Not really, I'm just used to writing match statements. Thats a good point though, I'll swap in an if/else.

**Comment by holdenk** [comment]
I believe the test failure is spurious (looks like one of the workers is off-line). Jenkins retest this please.

**Comment by holdenk** [comment]
re-ping @dbtsai

**Comment by holdenk** [comment]
seems unrelated, jenkins retest this please.

File: mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+          case _ => {
+            val nodeSplit = node.split
+            val (leftSubCode, leftSubFunction) = nodeToTree(node.leftChild, depth + 1)
+            val (rightSubCode, rightSubFunction) = nodeToTree(node.rightChild, depth + 1)
+            val subCode = nodeSplit match {
+              case split: CategoricalSplit => {
+                val isLeft = split.isLeft
+                isLeft match {
+                  case true => s"""
+                              if (categories.contains(input.apply(${split.featureIndex}))) {
```
Comment (frreiss) [inline mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala]: For small sets, this comparison really should be inlined. 
For double-valued columns, you can generate a series of equality comparisons against constants OR'd together. If you choose the structure of the comparisons carefully, the JIT compiler should generate SIMD instructions that do four comparisons per clock cycle.
For integer-encoded columns with less than 64 categories, you can generate a bitwise AND against a single long constant.

**Comment by frreiss** [comment]
LGTM as far as correctness goes. There's still quite a bit of pointer chasing going on in the generated code, though.

File: mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala (hunk 6 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala]: Sounds like a good idea, I'll take a look at in-lining this for small sets.

**Comment by holdenk** [comment]
Streaming failures are probably unrelated, jenkins retest this please.

**Comment by holdenk** [comment]
One open question, thats come from keeping up with master, for the reviewers (I guess @dbtsai once he has some spare cycles) is if code generator should live under tree or tree.impl.

**Comment by holdenk** [comment]
The failure seems unrelated (spilling keys elsewhere) - jenkins retest this please.

**Comment by holdenk** [comment]
ping @dbtsai ?

**Comment by holdenk** [comment]
@dbtsai what do you think?

File: mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala [hunk 7]
```diff
@@ -209,6 +210,14 @@ final class GBTRegressionModel private[ml](
   @Since("1.4.0")
   override def treeWeights: Array[Double] = _treeWeights
 
+  // We use two vals with options rather than lazy vals since we want the codeGen to be
+  // eagerly evaluated.
+  val (treePredictors, codeGenPredictors) = if (useCodeGen) {
```
Comment (rxin) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GBTRegressor.scala]: are these public apis?

**Comment by holdenk** [comment]
re-ping?

**Comment by holdenk** [comment]
Now that 2.0 is mostly out - does @dbtsai have some cycles to review it?

**Comment by holdenk** [comment]
ping @dbtsai if you have some cycles :)

**Comment by holdenk** [comment]
Hey so I know this PR has been languishing for awhile, I'd be happy to do some updates or run some more update performance numbers after 2.1 is shipped or whatever might be neededif one of the commiters has time to look at this (e.g. @dbtsai / @jkbradley / @MLnick ) after 2.1 - does anyone feel like they might have bandwidth here?

**Comment by holdenk** [comment]
So I'm going to go ahead and close this until @dbtsai (or someone) lets me know they have the review cycles to look at this - then I'm happy to re-open and update to master.
